### ISSUE TITLE
Theory Part 2 - Gradient Descent and Backprop

### ISSUE THREAD

**Issue by HuwWDay** [Bristol-Training/intro_to_AI#5]
Please *don't* start work on this until you've met with Huw!

Aim to finish these at latest Monday 28th April. Schedule a meeting halfway through April to touch base. Put all of the notebooks in the Theory Part 2 folder in the github.

We need to explain gradient descent and backpropagation in a brief but intuitive way. General idea is going to be:

Inspired by: https://www.youtube.com/watch?v=IHZwWFHWa-w&list=PLZHQObOWTQDNU6R1_67000Dx_ZCJB-3pi&index=2 (and the rest of the playlist to be honest)

![Image](https://github.com/user-attachments/assets/ce20bc3a-0af7-46ee-af23-16b6e8ea6781)

**Gradient Descent**

- [ ] Explain and show that your error (sum of squared residuals) is a function of your weights and biases
- [ ] Fix all weights and biases at some value (could make them all 0 or 1) except for w1 and then have a plot with values of w1 on the x-axis and the sum of squared residuals output on the y-axis and say "well we want to make the error smaller so we want to change our weights and biases to make it smaller"
- [ ] Explain that if we want to find the lowest point, we want to figure out where is uphill and where is downhill and head downhill.  We can do this using the partial derivative of the loss with respect to the weight which tells us how steep the surface is and in what direction.
- [ ] Then increase the complexity and have a 3D plot with two different weights on the x and y axis and then the z-axis is like a surface of the loss height. 
- [ ] Say something like: "we have loads of parameters so our loss surface is quite hard to imagine, but if we take the gradient of the loss function at a given point, we know which way is downhill and how steep it is to go"
- [ ] In practise we can choose how much we take a step in that direction by tuning a parameter called a **learning rate** and explain how you might want to vary this (could link to this blog: https://cs231n.github.io/neural-networks-3/#sanitycheck)
- [ ] ...or there are methods called stochastic gradient descent which might not take into account all the data at once but only look at how a random subset of the data affects the error.

**Back propagation**

- [ ] Basically say, now that we know the effect of our weights and biases on the loss, we can tweak the weights and biases in a way that reduces the error
- [ ] For more detail, look at the 3blue1brown video: https://www.youtube.com/watch?v=Ilg3gGewQ5U and https://www.youtube.com/watch?v=Ilg3gGewQ5U
- [ ] Last bit, allude to Stochastic Gradient Descent and ADAM (but don't explain them much)!

![Image](https://github.com/user-attachments/assets/27a74e0a-663e-4548-b412-949585878cf4)
![Image](https://github.com/user-attachments/assets/fd279507-db9a-4484-bd0e-75f7d24a0ad6)

This could be one or two notebooks at most I think!
